### DISCUSSION TITLE
CRUD--Update not working.

### DISCUSSION THREAD

**Discussion by mulewile** [neuefische/questions#589] [Web] [answered]
I am stuck trying to figure out why my update function does not work. When I click on update button I get the following error. 

**Console Error**
> Request URL: http://localhost:3000/api/events/undefined
Request Method: GET
Status Code: 500 Internal Server Error
Remote Address: [::1]:3000
Referrer Policy: strict-origin-when-cross-origin

**My update function.**
```jsx
async function handleUpdateEvents(editedEvent) {
    setIsEditing(false);

    {
      const response = await fetch(id ? `/api/events/${id}` : null, {
        method: "PATCH",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(editedEvent),
      });
      router.push(`/events/${id}`);
      console.log("id CHECK ====>", id);
      if (!response.ok) {
        console.error(`Error: ${response.status}`);
      }
      console.log("Event edited....");
    }
  }
  ```
  [Source code for full context](https://github.com/mulewile/capstone-project/pull/16)
  
  I have tried to resolve the problem with some cohort members, google but in vain.

**Comment by onemanwenttomow** [comment, marked as answer]
I think you need to look at the request url a bit closer. In particular the end of the url, which somehow has undefined in it. 

`http://localhost:3000/api/events/undefined`

The fetch request is made in `handleUpdateEvents` in `_app.js` and currently constructs the url as so.

``` js 
await fetch(id ? `/api/events/${id}`
``` 

At the moment id is `undefined`(coming from)

``` js 
const router = useRouter();
const { id } = router.query;
```

We actually have access to the id of the event from the object that is passed to the `handleUpdateEvents` function. 

e.g. `const eventId = editedEvent._id;`

So we could do something like this. 

``` js 
const eventId = editedEvent._id;

const response = await fetch(eventId ? `/api/events/${eventId}` : null, {
  method: "PATCH",
  headers: {
    "Content-Type": "application/json",
  },
  body: JSON.stringify(editedEvent),
});
```

### Other thoughts. 

I think we could move the `handleUpdateEvents` into the `Form` component (or at least into the `/events/edit.js` file and then pass it down as a prop to the form). That way there is no need to pass the function all the way down from `_app.js` as props.
